controllers/imagerepository_controller.go

```diff
@@ -0,0 +1,580 @@
+				log.Error(err, "failed to remove image repository finalizer", l.Action, l.ActionUpdate)
+				return ctrl.Result{}, err
+			}
+			log.Info("Image repository finalizer removed", l.Action, l.ActionDelete)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if imageRepository.Status.State == imagerepositoryv1beta1.ImageRepositoryStateFailed {
+		return ctrl.Result{}, nil
+	}
+
+	// Reread quay token
+	r.QuayClient = r.BuildQuayClient(log)
+
+	// Provision image repository if it hasn't been done yet
+	if !controllerutil.ContainsFinalizer(imageRepository, ImageRepositoryFinalizer) {
+		if err := r.ProvisionImageRepository(ctx, imageRepository); err != nil {
+			log.Error(err, "provision of image repository failed")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if imageRepository.Status.State != imagerepositoryv1beta1.ImageRepositoryStateReady {
+		return ctrl.Result{}, nil
+	}
+
+	// Make sure, that image repository name is the same as on creation.
+	// Do it here to avoid webhook creation.
+	imageRepositoryName := strings.TrimPrefix(imageRepository.Status.Image.URL, fmt.Sprintf("quay.io/%s/", r.QuayOrganization))
+	if imageRepository.Spec.Image.Name != imageRepositoryName {
+		oldName := imageRepository.Spec.Image.Name
+		imageRepository.Spec.Image.Name = imageRepositoryName
+		if err := r.Client.Update(ctx, imageRepository); err != nil {
+			log.Error(err, "failed to revert image repository name", "OldName", oldName, "ExpectedName", imageRepositoryName, l.Action, l.ActionUpdate)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Change image visibility if requested
+	if imageRepository.Spec.Image.Visibility != imageRepository.Status.Image.Visibility && imageRepository.Spec.Image.Visibility != "" {
+		if err := r.ChangeImageRepositoryVisibility(ctx, imageRepository); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Rotate credentials if requested
+	regenerateToken := imageRepository.Spec.Credentials.RegenerateToken
+	if regenerateToken != nil && *regenerateToken {
+		if err := r.RegenerateImageRepositoryCredentials(ctx, imageRepository); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// ProvisionImageRepository creates image repository, robot account(s) and secret(s) to acces the image repository.
+// If labels with Application and Component name are present, robot account with pull only access
+// will be created and pull token will be propagated to all environments via Remote Secret.
+func (r *ImageRepositoryReconciler) ProvisionImageRepository(ctx context.Context, imageRepository *imagerepositoryv1beta1.ImageRepository) error {
+	log := ctrllog.FromContext(ctx).WithName("ImageRepositoryProvision")
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	imageRepositoryName := ""
+	if imageRepository.Spec.Image.Name == "" {
+		imageRepositoryName = imageRepository.Namespace + "-" + imageRepository.Name
```

Comment:
would be possible to join them by `/`? It's more userfriendly quay.io/org/my-tenant/name

Reply:
Technically yes, but in my opinion `/` is like a separator between registry and image name.

Reply:
ok, switching to `/` as separator